endpoint/policy: Keep internals separate

jrajahalme opened the pull request:
Fixes due to redirects being now created before mapstate is computed:
- Remove listener from MapStateEntry, was only needed to generate proxyID for proxy port retrieval. Now proxy port is already available before MapStateEntries are created, so the listener name is no longer needed
- Remove placeholder 'unrealizedRedirectPort', which was only needed when new redirects were created after mapstate had already been computed

Fixes: #35350

Save the old entry value when overridden by deny. MapStateEntry merge logic was simplified by only calling merge() if both entries are allows or denies. Now the caller overrides an allow with deny by overwriting it, but that branch missed the storing of the old value.

Fixes: #34437

Keep policy MapState internals private to the policy package. Replace the `MapState` interface with a limited set of accessor functions on `EndpointPolicy`. This way the endpoint does need direct access to the MapState map with the policy package internal bookkeeping.

Do not update the realized policy from the endpoint package when applying map changes to bpf policy maps. This is possible since the realized policy is either the same as the desired policy (when applying incremental changes) or the pending desired policy from which the changes are being applied is going to replace the old realized policy when the endpoint regeneration succeeds. If the endpoint regeneration fails, we also need to have the prior realized policy intact as the "last known good policy" which can be restored against a fresh dump of the policy map.

This also fixes the problem where MapStateEntries inserted from the endpoint package were missing the internal bookkeeping, like "owners", "derivedFromRules", etc.

Simplify `denyPreferredInsertWithChanges` by updating mapstate directly when iterating and simplify insertion of dependent keys with helpers and by using normal `if/else` structure.

Removal of the `Listener` field together with structural changes to keep mapstate internals private speed up `BenchmarkRegenerateCIDRDenyPolicyRules` by ~17% (~25% less memory used with ~12% less allocations)

jrajahalme committed:
policy: Make ChangeState.Old private

Prepare for hiding internals of MapStateEntry by making ChangeState.Old
private.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Clean up test logging

Remove extraneous printing and Use t.Logf() instead of fmt.Printf().

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Test with DeepEquals

Change key tests to validate full MapStateEntries by changing from Equals
to DeepEquals.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Remove unnecessary actions

Remove the following actions in
TestMapState_denyPreferredInsertWithSubnets test left unnecessary since
the introduction of the transactional selector cache:

- insertAWithBProto
- insertAasB
- insertBWithAProto
- insertBWithAProtoAsDeny

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Insert mapstate updates directly

Upon reflecting on the iterators used when deleting covered entries, or
adding new ones due to precedence, it is safe to add them directly, as
the iterators hold enough state for it to be safe to delete entries on
descendants, and add entries to existing trie nodes.

Simplify insertion of dependent keys with new helpers.

Simplify the structure of 'denyPreferredInsertWithChanges' by not
returning early on the end of the deny block.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Update existing keys without checking for structural changes

When an existing entry is updated without changing its IsDeny flag the
update can be done on the 'entries' map without delete from the other
map, or without sanity checking the key.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Make MapStateEntry.DerivedFromRules private

Do not export DerivedFromRules member so that we can change the
implementation later. Remove from tests where it's value was ignored
anyway, and add actual validation for them in the endpoint integration
tests. Note that these were ignored there as well before, so there are
some missing labels, now in comments. To be fixed in forthcoming commits.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Save the old entry value when overridden by deny

Old value must be saved also when overriding an allow entry with a deny.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: delete unrealizedRedirectPort

Redirects are now created before EndpointPolicy, so all redirects already
exist when processing incremental changes for the EndpointPolicy. Hence
'unrealizedRedirectPort' symbol is no longer needed.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Keep internals private

Define mapStateEntry to hold the internal bookkeeping and contain
MapStateEntry without them. This way the containters used for internal
bookkeeping are not exposed to other packages.

Remove the MapState interface that was ill-defined due to mixing &
matching exported and non-exported methods. Within the policy package use
'*mapState' directly, and export the needed functions via
'EndpointPolicy'.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Remove listener name from MapStateEntry

Now that proxy redirects are created before mapstate entries, we no
longer need the listener name in the MapStateEntry.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
endpoint: remove unused 'incremental' attribute, fix reported deleted entry

Remove unused 'incremental' attribute from policy map updates.

Store the actual delete entry value rather than the missing empty value
in reported mapstate diffs.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -1072,58 +1064,66 @@ func (e *Endpoint) applyPolicyMapChanges(regenContext *regenerationContext, hasN
+func (e *Endpoint) syncRealizedPolicyMap() error {
+	errors := 0
+
+	// Add policy map entries before deleting to avoid transient drops
```
@nathanjsweet realized we need to handle the case where add-then-delete pushes us over the mapstate limit, even if the end-result will fit.

So, you can either compute the sizes beforehand, or detect an oversize error, and re-try adds after deletes in that case.

jrajahalme replied:
Added a commit on the latter strategy.

squeed commented on pkg/policy/resolve.go:
```diff
@@ -220,25 +208,57 @@ func (p *EndpointPolicy) Detach() {
-// map.
-func NewMapStateWithInsert() (MapState, func(k Key, e MapStateEntry)) {
-	currentMap := newMapState()
+func (p *EndpointPolicy) Updated(realized *EndpointPolicy) iter.Seq2[Key, MapStateEntry] {
```
A brief docblock on these three functions would be nice.

jrajahalme replied:
Added :-)

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -1072,58 +1064,66 @@ func (e *Endpoint) applyPolicyMapChanges(regenContext *regenerationContext, hasN
-// not modified.
-func (e *Endpoint) syncPolicyMapWith(realized policy.MapState, withDiffs bool) (diffCount int, diffs []policy.MapChange, err error) {
+// Only called when desired and realized policies are not the same.
+func (e *Endpoint) syncRealizedPolicyMap() error {
```
Is there a reason this is different that `syncPolicyMapWith(e.realized)`?

The naming and functionality between these two functions is a bit confusing; they take a purported base state ("realized"), a new state ("desired") and apply the diffs to BPF but not the base state.

My suggestion: create a function, `e.applyPolicyMap(realized, desired, withDiffs)` that does:

1. if `realized` is nil, then dumps the current BPF map state
2. determines delta between desired and realized
3. applies that delta, gathering diffs for debugging if interesting.

This should allow us to unify all uses of this.

jrajahalme replied:
The 'realized' map is of a different type between these functions. It is the bpf policy map that is being synced. As an immediate remedy to limit confusion I've renamed `syncRealizedPolicyMap` as `syncPolicyMap` (it is not the agent's notion of the realized map that is being synced, but the bpf policy map) and added docblocks.

jrajahalme replied:
The general strategy on revert is to use the prior `e.realizedPolicy` as the "last known good policy" (it is the state of the policy map before this regeneration started). `e.desiredPolicy` is set to `e.realizedPolicy` first, then a fresh dump of the policy map is taken, and the bpf policy map is synced using those two. After this the bpf policy map should have the entries in `e.desiredPolicy`.

jrajahalme committed:
endpoint: Retry adding policy map entries after deletes

We generally want to add new policy map entries to avoid transient
drops. Since policy map size is limited, it is possible that adds fail
due to hitting the size limit. In that case it is possible that more
space would be available if any entries are deleted, so it makes sense to
retry after key deletion.

Note that since we no longer update the realized map on successful map
updates, we need to do the full map diff again.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
endpoint: Skip updating realized mapstate when updating bpf policy map

Skip updating the realized mapstate when updating bpf policy map, as:

1. Usually the realized mapstate is the same as desired mapstate so there is
   nothing to update

2. The realized mapstate is ephemeral dump from the datapath, there is no
   point updating it.

3. When a new policy has been computed, the new desired mapstate will
   replace the old realized mapstate anyway after the successful policy
   update. If policy update fails use a full policy map dump for restoring
   to the last known good policy.

   Not updating the realized policy also keeps the old realized policy as
   "the last known good" policy that we can restore to when policy update
   fails.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

squeed commented on pkg/policy/mapstate.go:
```diff
@@ -972,48 +1004,90 @@ func (ms *mapState) revertChanges(changes ChangeState) {
+// L3-only deny entries over L3-L4 allows.
+//
+// Since bpf datapath denies by default, we only need to add deny entries to carve out more specific
+// holes to less specific allow rules. But since we don't if allow entries will be added later
```
nit: "since we don't know", presumably :-)
